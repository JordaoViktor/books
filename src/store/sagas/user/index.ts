import { api } from "@services/api";
import { all, call, put, select, takeEvery, takeLatest } from "redux-saga/effects";

import { loginFailure, loginSuccess, login } from '@store/slices/user/index'
import { Alert } from "react-native";
import AsyncStorage from "@react-native-async-storage/async-storage";

interface IUserSignInAuth {
  payload: {
    email: string;
    password: string;
  }
}

export function* loginSaga({ payload: { email, password } }: IUserSignInAuth) {
  try {
    const { data, headers } = yield call(
      api.post,
      '/auth/sign-in',
      { email, password }
    );

    const user = { ...data, token: headers.authorization };

    console.log('Im a user:', headers)
    api.interceptors.request.use((config: any) => {

      config.headers.authorization = `Bearer ${user.token}`
      config.headers['refresh-token'] = headers['refresh-token']
      config.headers['Content-Type'] = 'application/json';

      return config;
    });

    yield call(AsyncStorage.setItem, '@user', JSON.stringify(user));
    yield put(loginSuccess(data));

  } catch (error) {
    if (error instanceof Error) {
      yield put(
        loginFailure({
          error: true,
          status: error.response.status
        })
      );
    }
  }
}

export default function* listener() {
  yield all([
    takeLatest(login().type, loginSaga),
  ])
}