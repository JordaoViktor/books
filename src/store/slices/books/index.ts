import { createSlice } from '@reduxjs/toolkit'

const initialState = {
  data: null,
  error: null,
  statusCode: null,
  isLoading: false,
}

const bookSlice = createSlice({
  name: '@books',
  initialState,
  reducers: {
    fetchBooks(draft, { payload }) {
      draft.error = null;
      draft.isLoading = true;
    },
    fetchBooksSuccess(draft, { payload }) {
      draft.data = payload.data;
      draft.isLoading = false;
    },
    fetchBooksFailure(draft, { payload }) {
      draft.error = payload.error;
      draft.isLoading = false;
      draft.statusCode = payload.status;
    },
  }
})

export const { fetchBooks, fetchBooksSuccess, fetchBooksFailure } = bookSlice.actions

export const bookReducer = bookSlice.reducer