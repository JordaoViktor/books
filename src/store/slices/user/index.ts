import { createSlice } from '@reduxjs/toolkit'

const initialState = {
  data: null,
  error: null,
  statusCode: null,
  isLoading: false,
}

const userSlice = createSlice({
  name: '@user',
  initialState,
  reducers: {
    login(draft, { payload }) {
      draft.error = null;
      draft.data = payload.data;
      draft.isLoading = true;
    },
    loginSuccess(draft, { payload }) {
      draft.data = payload.data;
      draft.isLoading = false;
    },
    loginFailure(draft, { payload }) {
      draft.error = payload.error;
      draft.isLoading = false;
      draft.statusCode = payload.status;
    },
  }
})

export const { login, loginSuccess, loginFailure } = userSlice.actions

export const userReducer = userSlice.reducer