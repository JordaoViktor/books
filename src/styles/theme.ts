export default {
  colors: {
    primary: "#FFFF",
    secondary: "#333333",
    tertiary: "#AB2680",
    background: "#E5E5E5",
    dark: "#000",

    lightenText: "#FFFF",
    darkText: "#333333",

    darkOpacity200: "rgba(0,0,0,0.20)",
    darkOpacity300: "rgba(0,0,0,0.32)",
    darkOpacity400: "rgba(0,0,0,0.4)",

    transparent: "transparent"
  },
  fonts: {
    light: "Heebo_300Light",
    normal: "Heebo_400Regular",
    bold: "Heebo_500Medium",
  },
}