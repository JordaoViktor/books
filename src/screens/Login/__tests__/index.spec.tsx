import React from 'react';
import renderer from 'react-test-renderer';
import { render, cleanup } from '@testing-library/react-native'
import { ThemeProvider } from 'styled-components/native'
import theme from '@styles/theme'
import { Login } from '../index';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store'
import { store } from '@store/index'
import createSagaMiddleware from 'redux-saga';
import mockAxios from 'jest-mock-axios';
import 'jest-styled-components'
import { login } from '@store/slices/user';
interface IProviderProps {
  children?: React.ReactNode;
  store?: any;
}


const ReduxProvider = ({ children }: IProviderProps) => {
  const initialState = {
    data: null,
    error: null,
    statusCode: null,
    isLoading: false
  }
  const sagaMiddleware = createSagaMiddleware();

  const mockStore = configureStore([sagaMiddleware])(initialState)

  return (
    <Provider store={mockStore} >
      <ThemeProvider theme={theme} >
        {children}
      </ThemeProvider>
    </Provider>
  )
};

test('matches login screen snapshot', async () => {

  afterEach(() => {
    // cleaning up the mess left behind the previous test
    mockAxios.reset();
  });
  afterAll(() => { jest.useFakeTimers() })
  afterEach(cleanup)

  // const { getByTestId, debug } = render(<Login />, { wrapper: ReduxProvider })
  // const screenRender = getByTestId('login-screen')

  // console.log('screen render: ', screenRender)
  // console.log(debug)

  const tree = renderer
    .create(
      <ReduxProvider >
        <Login />
      </ReduxProvider>
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
  // const FormInputRender = renderer
  //   .create(
  //     <Login />
  //     , { wrapper: ()=> ReduxProvider })
  //   .toJSON();

  // await expect(FormInputRender).toMatchSnapshot();
});