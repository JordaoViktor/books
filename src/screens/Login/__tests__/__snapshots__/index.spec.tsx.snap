// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Screen matches login screen snapshot 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "justifyContent": "center",
      },
    ]
  }
>
  <Image
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    resizeMode="cover"
    source={1}
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": undefined,
          "width": undefined,
        },
        undefined,
      ]
    }
    testID="login-screen"
  />
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Array [
        Object {
          "height": 280,
          "width": "90%",
        },
      ]
    }
  >
    <View
      accessibilityLabel="Ioasys Books"
      accessible={true}
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 40,
            "width": "100%",
          },
        ]
      }
    >
      SvgMock
      <Text
        style={
          Array [
            Object {
              "color": "#FFFF",
              "fontFamily": "Heebo_300Light",
              "fontSize": 55,
              "marginLeft": 16,
            },
          ]
        }
      >
        Books
      </Text>
    </View>
    <View
      marginTop={50}
      style={
        Array [
          Object {
            "backgroundColor": "rgba(0,0,0,0.32)",
            "borderRadius": 4,
            "marginTop": 50,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#FFFF",
              "fontFamily": "Heebo_400Regular",
              "fontSize": 24,
              "marginBottom": 5,
              "opacity": 0.5,
              "paddingBottom": 0,
              "paddingLeft": 15,
              "paddingRight": 0,
              "paddingTop": 8,
              "position": "absolute",
            },
          ]
        }
      >
        Email
      </Text>
      <View
        style={
          Array [
            Object {
              "marginLeft": 50,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#FFFF",
                "fontFamily": "Heebo_400Regular",
                "fontSize": 24,
                "marginBottom": 5,
                "opacity": 0.5,
                "paddingBottom": 0,
                "paddingLeft": 15,
                "paddingRight": 0,
                "paddingTop": 8,
                "position": "absolute",
              },
            ]
          }
        />
      </View>
      <View
        style={
          Array [
            Object {
              "flexDirection": "row",
              "height": 60,
              "opacity": 1,
              "paddingLeft": 15,
              "paddingTop": 15,
              "width": "100%",
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "width": "100%",
              },
            ]
          }
        >
          <TextInput
            accessibilityLabel="Email address input"
            accessible={true}
            allowFontScaling={true}
            autoCapitalize="none"
            keyboardType="email-address"
            onChangeText={[Function]}
            rejectResponderTermination={true}
            style={
              Array [
                Object {
                  "color": "#FFFF",
                  "fontFamily": "Heebo_400Regular",
                  "fontSize": 31,
                  "height": "100%",
                  "width": "100%",
                  "zIndex": 3,
                },
              ]
            }
            testID="input"
            underlineColorAndroid="transparent"
          />
        </View>
      </View>
    </View>
    <View
      marginTop={16}
      style={
        Array [
          Object {
            "backgroundColor": "rgba(0,0,0,0.32)",
            "borderRadius": 4,
            "marginTop": 16,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#FFFF",
              "fontFamily": "Heebo_400Regular",
              "fontSize": 24,
              "marginBottom": 5,
              "opacity": 0.5,
              "paddingBottom": 0,
              "paddingLeft": 15,
              "paddingRight": 0,
              "paddingTop": 8,
              "position": "absolute",
            },
          ]
        }
      >
        Senha
      </Text>
      <View
        style={
          Array [
            Object {
              "marginLeft": 50,
            },
          ]
        }
      >
        <Text
          accessibilityValue={
            Object {
              "text": "undefined",
            }
          }
          accessible={true}
          style={
            Array [
              Object {
                "color": "#FFFF",
                "fontFamily": "Heebo_400Regular",
                "fontSize": 24,
                "marginBottom": 5,
                "opacity": 0.5,
                "paddingBottom": 0,
                "paddingLeft": 15,
                "paddingRight": 0,
                "paddingTop": 8,
                "position": "absolute",
              },
            ]
          }
        />
      </View>
      <View
        style={
          Array [
            Object {
              "flexDirection": "row",
              "height": 60,
              "opacity": 1,
              "paddingLeft": 15,
              "paddingTop": 15,
              "width": "100%",
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "width": "100%",
              },
            ]
          }
        >
          <TextInput
            accessibilityLabel="Password input"
            allowFontScaling={true}
            autoCorrect={false}
            onChangeText={[Function]}
            rejectResponderTermination={true}
            secureTextEntry={true}
            style={
              Array [
                Object {
                  "color": "#FFFF",
                  "fontFamily": "Heebo_400Regular",
                  "fontSize": 31,
                  "height": "100%",
                  "width": "100%",
                  "zIndex": 3,
                },
              ]
            }
            testID="input"
            underlineColorAndroid="transparent"
          />
        </View>
        <View
          style={
            Array [
              Object {
                "marginLeft": -100,
                "width": 85,
                "zIndex": 4,
              },
            ]
          }
        >
          <View
            accessibilityLabel="Sign in Button"
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#FFFF",
                "borderColor": "transparent",
                "borderRadius": 20,
                "borderWidth": 1,
                "height": 36,
                "justifyContent": "center",
                "opacity": 1,
                "paddingBottom": 9,
                "paddingLeft": 9,
                "paddingRight": 9,
                "paddingTop": 9,
              }
            }
          >
            <Text
              color="#AB2680"
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "color": "#AB2680",
                    "fontFamily": "Heebo_500Medium",
                    "fontSize": 31,
                    "justifyContent": "center",
                    "marginTop": -2,
                    "textAlign": "center",
                  },
                ]
              }
            >
              Entrar
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTSafeAreaView>
</View>
`;
